### PR TITLE
ssbc: Fix double dequeues after worker fairness PR

### PR DESCRIPTION
**Pull request by eseliger**
There was one problem with the solution that I didn't see while testing locally, because it doesn't happen when only 1 executor is attached: We dequeue the same record multiple times.
This is very weird to me, as I would've thought the DB should error here instead, but that's not the case.
What happens: With our complex view query that contains a materialized CTE in the middle of the chain, postgres loses track of the underlying table and doesn't lock the record on there, but on the materialized CTE (which is .. noop?). Hence, two executors can briefly dequeue the same record. 
I found this to be a larger problem with arbitrary views, so I made the fix in the dbworker implementation instead.
Also, this now has a second view that we actually pull the data from to include the other, non-queued records as well. This is needed, because the executor scaling metric needs to be able to view processing records to do proper scaling. This was also broken, but is harder to notice.
This fixes it by first selecting from the view and then properly reselecting from the underlying table. The query sounds slow, but actually it's quite fine. The query plan looks horrible I admit, but the runtime is really good. This has been tested on a DB with a few hundred thousand rows:

```sql
explain WITH potential_candidates AS (
	SELECT
		id AS candidate_id,
		ROW_NUMBER() OVER (ORDER BY batch_spec_workspace_execution_jobs.place_in_global_queue) AS order
	FROM batch_spec_workspace_execution_jobs_with_rank batch_spec_workspace_execution_jobs
	WHERE
		(
			(
				state = 'queued' AND
				(process_after IS NULL OR process_after <= NOW())
			) OR (
				0 > 0 AND
				state = 'errored' AND
				NOW() - finished_at > (30 * '1 second'::interval) AND
				num_failures < 5
			)
		)
	ORDER BY batch_spec_workspace_execution_jobs.place_in_global_queue
),
candidate AS (
	SELECT
		id FROM batch_spec_workspace_execution_jobs
	JOIN potential_candidates pc ON pc.candidate_id = id
	WHERE
		-- Recheck state.
		state = 'queued'
	ORDER BY pc.order
	FOR UPDATE OF batch_spec_workspace_execution_jobs SKIP LOCKED
	LIMIT 1
),
updated_record AS (
	UPDATE
		batch_spec_workspace_execution_jobs
	SET
		state = 'processing'
	WHERE
		id IN (SELECT id FROM candidate)
)
SELECT
	batch_spec_workspace_execution_jobs.id,
	batch_spec_workspace_execution_jobs.place_in_global_queue,
	batch_spec_workspace_execution_jobs.place_in_user_queue
FROM
	batch_spec_workspace_execution_jobs_with_rank batch_spec_workspace_execution_jobs
WHERE
	id IN (SELECT id FROM candidate);
```

```
                                                                                                                           QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop Left Join  (cost=36899.15..36900.33 rows=1 width=24) (actual time=0.078..0.081 rows=0 loops=1)
   Join Filter: (j.id = queue_candidates.id)
   CTE candidate
     ->  Limit  (cost=18450.27..18450.29 rows=1 width=62) (actual time=0.076..0.078 rows=0 loops=1)
           ->  LockRows  (cost=18450.27..18450.29 rows=1 width=62) (actual time=0.076..0.077 rows=0 loops=1)
                 ->  Sort  (cost=18450.27..18450.28 rows=1 width=62) (actual time=0.075..0.077 rows=0 loops=1)
                       Sort Key: pc."order"
                       Sort Method: quicksort  Memory: 25kB
                       ->  Nested Loop  (cost=18448.44..18450.26 rows=1 width=62) (actual time=0.034..0.036 rows=0 loops=1)
                             Join Filter: (batch_spec_workspace_execution_jobs.id = pc.candidate_id)
                             ->  Index Scan using batch_spec_workspace_execution_jobs_state on batch_spec_workspace_execution_jobs  (cost=0.42..2.21 rows=1 width=14) (actual time=0.034..0.034 rows=0 loops=1)
                                   Index Cond: (state = 'queued'::text)
                             ->  Subquery Scan on pc  (cost=18448.01..18448.04 rows=1 width=56) (never executed)
                                   ->  WindowAgg  (cost=18448.01..18448.03 rows=1 width=24) (never executed)
                                         ->  Sort  (cost=18448.01..18448.02 rows=1 width=16) (never executed)
                                               Sort Key: (row_number() OVER (?))
                                               ->  Nested Loop Left Join  (cost=18446.17..18448.00 rows=1 width=16) (never executed)
                                                     Join Filter: (j_1.id = queue_candidates_1.id)
                                                     ->  Index Scan using batch_spec_workspace_execution_jobs_state on batch_spec_workspace_execution_jobs j_1  (cost=0.42..2.21 rows=1 width=8) (never executed)
                                                           Index Cond: (state = 'queued'::text)
                                                           Filter: ((process_after IS NULL) OR (process_after <= now()))
                                                     ->  WindowAgg  (cost=18445.74..18445.77 rows=1 width=24) (never executed)
                                                           ->  Subquery Scan on queue_candidates_1  (cost=18445.74..18445.76 rows=1 width=8) (never executed)
                                                                 ->  Sort  (cost=18445.74..18445.75 rows=1 width=36) (never executed)
                                                                       Sort Key: (rank() OVER (?)), (max(exec_3.started_at)) NULLS FIRST
                                                                       ->  WindowAgg  (cost=18445.71..18445.73 rows=1 width=36) (never executed)
                                                                             ->  Sort  (cost=18445.71..18445.71 rows=1 width=28) (never executed)
                                                                                   Sort Key: exec_3.user_id, exec_2.created_at, exec_2.id
                                                                                   ->  Hash Join  (cost=18445.31..18445.70 rows=1 width=28) (never executed)
                                                                                         Hash Cond: (exec_3.user_id = exec_2.user_id)
                                                                                         ->  HashAggregate  (cost=18443.09..18443.25 rows=16 width=12) (never executed)
                                                                                               Group Key: exec_3.user_id
                                                                                               ->  Seq Scan on batch_spec_workspace_execution_jobs exec_3  (cost=0.00..16992.06 rows=290206 width=12) (never executed)
                                                                                         ->  Hash  (cost=2.21..2.21 rows=1 width=20) (never executed)
                                                                                               ->  Index Scan using batch_spec_workspace_execution_jobs_state on batch_spec_workspace_execution_jobs exec_2  (cost=0.42..2.21 rows=1 width=20) (never executed)
                                                                                                     Index Cond: (state = 'queued'::text)
   CTE updated_record
     ->  Update on batch_spec_workspace_execution_jobs batch_spec_workspace_execution_jobs_1  (cost=0.45..2.67 rows=1 width=285) (actual time=0.001..0.001 rows=0 loops=1)
           ->  Nested Loop  (cost=0.45..2.67 rows=1 width=285) (actual time=0.001..0.001 rows=0 loops=1)
                 ->  HashAggregate  (cost=0.02..0.03 rows=1 width=40) (actual time=0.000..0.000 rows=0 loops=1)
                       Group Key: candidate_1.id
                       ->  CTE Scan on candidate candidate_1  (cost=0.00..0.02 rows=1 width=40) (actual time=0.000..0.000 rows=0 loops=1)
                 ->  Index Scan using batch_spec_workspace_execution_jobs_pkey on batch_spec_workspace_execution_jobs batch_spec_workspace_execution_jobs_1  (cost=0.42..2.64 rows=1 width=217) (never executed)
                       Index Cond: (id = candidate_1.id)
   ->  Nested Loop  (cost=0.45..1.57 rows=1 width=8) (actual time=0.077..0.078 rows=0 loops=1)
         ->  HashAggregate  (cost=0.02..0.03 rows=1 width=8) (actual time=0.077..0.077 rows=0 loops=1)
               Group Key: candidate.id
               ->  CTE Scan on candidate  (cost=0.00..0.02 rows=1 width=8) (actual time=0.077..0.077 rows=0 loops=1)
         ->  Index Only Scan using batch_spec_workspace_execution_jobs_pkey on batch_spec_workspace_execution_jobs j  (cost=0.42..1.54 rows=1 width=8) (never executed)
               Index Cond: (id = candidate.id)
               Heap Fetches: 0
   ->  WindowAgg  (cost=18445.74..18445.77 rows=1 width=24) (never executed)
         ->  Subquery Scan on queue_candidates  (cost=18445.74..18445.76 rows=1 width=16) (never executed)
               ->  Sort  (cost=18445.74..18445.75 rows=1 width=36) (never executed)
                     Sort Key: (rank() OVER (?)), (max(exec_1.started_at)) NULLS FIRST
                     ->  WindowAgg  (cost=18445.71..18445.73 rows=1 width=36) (never executed)
                           ->  Sort  (cost=18445.71..18445.71 rows=1 width=28) (never executed)
                                 Sort Key: exec_1.user_id, exec.created_at, exec.id
                                 ->  Hash Join  (cost=18445.31..18445.70 rows=1 width=28) (never executed)
                                       Hash Cond: (exec_1.user_id = exec.user_id)
                                       ->  HashAggregate  (cost=18443.09..18443.25 rows=16 width=12) (never executed)
                                             Group Key: exec_1.user_id
                                             ->  Seq Scan on batch_spec_workspace_execution_jobs exec_1  (cost=0.00..16992.06 rows=290206 width=12) (never executed)
                                       ->  Hash  (cost=2.21..2.21 rows=1 width=20) (never executed)
                                             ->  Index Scan using batch_spec_workspace_execution_jobs_state on batch_spec_workspace_execution_jobs exec  (cost=0.42..2.21 rows=1 width=20) (never executed)
                                                   Index Cond: (state = 'queued'::text)
 Planning Time: 1.228 ms
 Execution Time: 0.432 ms
(68 rows)
```

Closes https://github.com/sourcegraph/sourcegraph/issues/37071

## Test plan

Adjusted test suites, wrote a test that currently fails on main but doesn't on this PR and made sure things still seem to work.

### CODE REVIEW

File: internal/workerutil/dbworker/store/store.go [hunk 1]
```diff
@@ -561,7 +575,27 @@ WITH candidate AS (
...
 		%s
 	ORDER BY %s
-	FOR UPDATE SKIP LOCKED
+),
+candidate AS (
+	SELECT
+		{id} FROM %s
+	JOIN potential_candidates pc ON pc.candidate_id = {id}
+	WHERE
+		-- Recheck conditions.
```
Comment (eseliger) [inline internal/workerutil/dbworker/store/store.go:584]: TODO: queue status should be enough

File: doc/dev/background-information/workers.md [hunk 2]
```diff
@@ -384,3 +384,7 @@ func makeResetter(workerStore store.Store) {
 The results of `makeWorker` and `makeResetter` can then be passed to `goroutine.MonitorBackgroundRoutines`.
 
 The worker and resetter may or or may execute in the same process. For example, we run all code intelligence background routines in the frontend, except for our LSIF conversion worker, which runs in a separate process for resource isolation and independent scaling.
+
+#### Step 7: Consider adding indexes
+
+The worker depends on a few columns to dequeue records. To keep it fast, consider adding indexes on the `state` and `process_after` columns.
```
Comment (eseliger) [inline doc/dev/background-information/workers.md:390]: I think we should also check at some point that most stores have these indexes. I suspect multiple workers are missing them.

File: doc/dev/background-information/workers.md [hunk 3]
```diff
@@ -295,19 +295,19 @@ import (
 func makeStore(db dbutil.DB) store.Store {
 	return store.New(db, store.Options{
 		Name:              "example_job_worker_store",
-		TableName:         "example_jobs j",
-		ViewName:          "example_jobs_with_repository_name j",
+		TableName:         "example_jobs",
+		ViewName:          "example_jobs_with_repository_name example_jobs",
```
Comment (eseliger) [inline doc/dev/background-information/workers.md:299]: I removed this behavior as it was very implicitly working and mucho hard to get to work with the new dequeue query.

File: internal/workerutil/dbworker/store/store.go [hunk 4]
```diff
@@ -561,7 +567,16 @@ WITH candidate AS (
...
+),
+candidate AS (
+	SELECT
+		{id} FROM %s
+	JOIN potential_candidates pc ON pc.candidate_id = {id}
+	WHERE
+		-- Recheck state.
+		{state} IN ('queued', 'errored')
+	ORDER BY pc.order
+	FOR UPDATE OF %s SKIP LOCKED
```
Comment (eseliger) [inline internal/workerutil/dbworker/store/store.go:579]: This time very explicit :D

Comment (eseliger) [reply internal/workerutil/dbworker/store/store.go:579]: Might fix some locking issues with joined tables, too.

File: doc/dev/background-information/workers.md [hunk 5]
```diff
@@ -177,7 +177,7 @@ We assume that the repository name is be necessary to process the record, meanin
 
 Next, we define the struct instance `ExampleJob` that mirrors the interesting fields of the `example_jobs_with_repository_name` view.
 
-We will additionally define an array of SQL column expressions that correspond to each field of the struct. For these expressions to be valid, we assume they will be embeddded in a query where `j` corresponds to a row of the `example_jobs_with_repository_name` table. Note that these expressions can be arbitrarily complex (conditional, sub-select expressions, etc).
+We will additionally define an array of SQL column expressions that correspond to each field of the struct. For these expressions to be valid, we assume they will be embeddded in a query where the `example_jobs` corresponds to a row of the `example_jobs_with_repository_name` table. Note that these expressions can be arbitrarily complex (conditional, sub-select expressions, etc).
```
Comment (LawnGnome) [inline doc/dev/background-information/workers.md:180]:
```suggestion
We will additionally define an array of SQL column expressions that correspond to each field of the struct. For these expressions to be valid, we assume they will be embeddded in a query where the `example_jobs` record corresponds to a row of the `example_jobs_with_repository_name` table. Note that these expressions can be arbitrarily complex (conditional, sub-select expressions, etc).
```

File: doc/dev/background-information/workers.md [hunk 6]
```diff
@@ -87,17 +87,17 @@ The store relies on a jobs table _specific to your worker_ to exist with the fol
...
+| `state`             | text                     | The job's current status (one of `queued`, `processing`, `completed`, `errored`, or `failed`) |
 | `failure_message`   | text                     | Updated with the text of the error returned from the handle hook |
-| `queued_at`         | timestamp with time zone | Time when the job was added to the table
+| `queued_at`         | timestamp with time zone | Time when the job was added to the table |
 | `started_at`        | timestamp with time zone | Updated when the job is dequeued for processing |
 | `finished_at`       | timestamp with time zone | Updated when the handler finishes processing the job (successfully or unsuccessfully) |
 | `process_after`     | timestamp with time zone | Controls the time after which the job is visible for processing |
 | `num_resets`        | integer                  | Updated when the job is moved back from `failed` to `queued` |
-| `num_failures`      | integer                  | Updated when the job enters the `failed` state |
+| `num_failures`      | integer                  | Updated when the job enters the `errored` state |
```
Comment (LawnGnome) [inline doc/dev/background-information/workers.md:97]: 😬

**Review by LawnGnome — APPROVED** [review]
LGTM. I have a slight bit of queasiness about how non-deterministic goroutine based tests can be, but I don't see an obvious alternative here.

**Comment by coury-clark** [comment]
@eseliger We're seeing very long queries that are hitting timeout in the past few days: https://k8s.sgdev.org/-/debug/grafana/goto/ahO8JQC7z.

Maybe the scale of this table (and therefore the index) causes this new query to break quite badly in ways that wasn't obvious from the testing above?

**Pull request merged**